File: python/day_18.py
with open('./inputs/18.txt', 'r') as infile:
    first_row = infile.read()


def count_safes(row, total_lines):
    safe = 0
    for i in range(total_lines):
        if i == 40:
            first_solution = safe
        safe += row.count('.')
        old = '.' + row + '.'
        row = ''
        for left, right in zip(old, old[2:]):
            row += '^' if left != right else '.'
    return first_solution, safe


first, second = count_safes(first_row, 400000)

print("Oh, so many traps here! Let me make 40 steps and count safe tiles.")
print(f"There are about {first} tiles here.")
print("....")
print("This room is 400,000 steps long?! What a large room!")
print(f"And there are {second} safe tiles in total.")

File: clojure/day18.clj
(ns day18
  (:require aoc))


(defn next-row [row]
  (into [(second row)]
        (map (fn [left right]
               (if (= left right) 1 0))
             row
             (drop 2 (conj row 1)))))

(defn make-tiles [puzzle iters]
  (first
   (reduce
    (fn [[cnt row] _]
      [(+ cnt (apply + row))
       (next-row row)])
    [0 puzzle]
    (range iters))))


(defn parse-char [c]
  (if (= c \.) 1 0))

(defn solve [input]
  (let [puzzle (->> (aoc/read-input-line input :chars)
                    (mapv parse-char))]
    (for [iters [40 400000]]
      (make-tiles puzzle iters))))


(solve 18)
